File: Number of paths[22-Oct-2023]/Solution.java
//   >> JAVA CODE <<

public class Solution {
    public int numberOfPaths(int M, int N) {
        final int mod = 1000000007;

        int n = M + N - 2;
        int r = M - 1;
        long ans = 1;

        // Calculate (n choose r) % mod
        for (int i = 1; i <= r; i++) {
            ans = (ans * (n - i + 1)) % mod;
            ans = (ans * modInverse(i, mod)) % mod;
        }

        return (int) ans;
    }

    private long modInverse(long a, int m) {
        return modPow(a, m - 2, m);
    }

    private long modPow(long base, int exp, int mod) {
        long result = 1;

        while (exp > 0) {
            if (exp % 2 == 1) {
                result = (result * base) % mod;
            }
            base = (base * base) % mod;
            exp /= 2;
        }

        return result;
    }
}

File: Number of paths[22-Oct-2023]/solution.py
#   >> PYTHON CODE <<

class Solution:
    def numberOfPaths(self, M, N):
        mod = int(1e9 + 7)
        
        n = M + N - 2
        r = M - 1
        ans = 1

        # Calculate (n choose r) % mod
        for i in range(1, r + 1):
            ans = (ans * (n - i + 1)) % mod
            ans = (ans * pow(i, mod - 2, mod)) % mod  # Calculate the modular inverse and use it for division

        return int(ans)